File: runtime/projection.go
package runtime

import (
	"errors"
	"fmt"
)

// ProjectView builds the public snapshot from the ordered public event stream.
// Operational indexes are deliberately absent from this projection.
func ProjectView(conversation Conversation, events []Event) (ConversationView, error) {
	view := ConversationView{Conversation: conversation}
	submissions := make(map[string]int)
	type toolKey struct{ runID, callID string }
	tools := make(map[toolKey]int)

	for _, event := range events {
		if event.ConversationID != conversation.ID {
			return ConversationView{}, errors.New("runtime: event belongs to another conversation")
		}
		if event.ID > view.EventCursor {
			view.EventCursor = event.ID
		}

		switch event.Type {
		case EventInputAccepted, EventAssistantCommitted, EventToolOutcomeRecorded:
			message, err := event.ProjectMessage()
			if err != nil {
				return ConversationView{}, err
			}
			view.Messages = append(view.Messages, message)
			switch event.Type {
			case EventInputAccepted:
				if event.Input.IdempotencyKey == "" {
					return ConversationView{}, errors.New("runtime: accepted input has no idempotency key")
				}
				if _, exists := submissions[event.InboundMessageID]; exists {
					return ConversationView{}, fmt.Errorf("runtime: input %q accepted twice", event.InboundMessageID)
				}
				submissions[event.InboundMessageID] = len(view.Submissions)
				view.Submissions = append(view.Submissions, Submission{
					ID:             event.InboundMessageID,
					ConversationID: conversation.ID,
					MessageID:      event.InboundMessageID,
					AgentRevision:  event.Input.AgentRevision,
					Status:         RunQueued,
					AcceptedAt:     event.CreatedAt,
				})
			case EventAssistantCommitted:
				for _, part := range event.AssistantOutput.Parts {
					if part.Type != "tool_call" {
						continue
					}
					key := toolKey{event.RunID, part.ToolCallID}
					if part.ToolCallID == "" || event.RunID == "" {
						return ConversationView{}, errors.New("runtime: tool request has no identity")
					}
					if _, exists := tools[key]; exists {
						return ConversationView{}, fmt.Errorf("runtime: tool call %q recorded twice", part.ToolCallID)
					}
					tools[key] = len(view.ToolCalls)
					view.ToolCalls = append(view.ToolCalls, ToolCall{
						ID:               part.ToolCallID,
						ConversationID:   conversation.ID,
						RunID:            event.RunID,
						InboundMessageID: event.InboundMessageID,
						Kind:             part.ToolKind,
						Arguments:        append([]byte(nil), part.Arguments...),
						Status:           ToolRequested,
						UpdatedAt:        event.CreatedAt,
					})
				}
			case EventToolOutcomeRecorded:
				key := toolKey{event.RunID, event.ToolOutcome.ToolCallID}
				index, ok := tools[key]
				if !ok {
					return ConversationView{}, fmt.Errorf("runtime: tool outcome %q has no request", key.callID)
				}
				view.ToolCalls[index].Status = event.ToolOutcome.Status
				view.ToolCalls[index].Result = event.ToolOutcome.Result
				view.ToolCalls[index].UpdatedAt = event.CreatedAt
			}
		case EventRunStarted, EventRunCompleted, EventRunFailed:
			if event.Run == nil {
				return ConversationView{}, errors.New("runtime: run event has no run")
			}
			index, ok := submissions[event.Run.InboundMessageID]
			if !ok {
				return ConversationView{}, fmt.Errorf("runtime: run %q has no accepted input", event.Run.ID)
			}
			view.Submissions[index].Status = event.Run.Status
			view.Submissions[index].Error = event.Run.Error
			if event.Type == EventRunStarted {
				view.ActiveRun = event.Run
			} else {
				view.ActiveRun = nil
			}
		case EventEnvironmentProgress:
			if event.EnvironmentProgress != nil {
				view.EnvironmentEvents = append(view.EnvironmentEvents, event)
			}
		}
	}
	return view, nil
}

File: web/src/projection.ts
import type { ConversationView, RuntimeEvent, ToolCall } from "./types";

// These match the runtime's NoticeFailedText and NoticeStoppedText so live
// replay renders the same message as a snapshot.
const noticeFailedText = "An error occurred while generating a response. Try again.";
const noticeStoppedText = "Stopped. The response was not finished.";

function upsert<T extends { id: string }>(items: T[], item: T): T[] {
  const index = items.findIndex((existing) => existing.id === item.id);
  if (index < 0) return [...items, item];
  const next = [...items];
  next[index] = item;
  return next;
}

function upsertToolCall(items: ToolCall[], item: ToolCall): ToolCall[] {
  const index = items.findIndex((existing) => existing.run_id === item.run_id && existing.id === item.id);
  if (index < 0) return [...items, item];
  const next = [...items];
  next[index] = item;
  return next;
}

export function applyEvent(view: ConversationView, event: RuntimeEvent): ConversationView {
  const next: ConversationView = {
    ...view,
    messages: view.messages,
    submissions: view.submissions,
    tool_calls: view.tool_calls,
    environment_events: view.environment_events,
    active_run: view.active_run,
  };
  if (event.cursor && event.cursor > next.event_cursor) next.event_cursor = event.cursor;

  switch (event.type) {
    case "input.accepted":
      if (event.input && event.inbound_message_id) {
        next.messages = upsert(next.messages, {
          id: event.inbound_message_id, conversation_id: event.conversation_id,
          inbound_message_id: event.inbound_message_id, role: "user", complete: true,
          parts: event.input.parts, created_at: event.created_at,
        });
        next.submissions = upsert(next.submissions ?? [], {
          id: event.inbound_message_id, conversation_id: event.conversation_id,
          message_id: event.inbound_message_id, agent_revision: event.input.agent_revision,
          status: "queued", accepted_at: event.created_at,
        });
      }
      break;
    case "assistant.output.committed":
      if (event.assistant_output) {
        next.messages = upsert(next.messages, {
          id: event.assistant_output.id, conversation_id: event.conversation_id,
          inbound_message_id: event.inbound_message_id, run_id: event.run_id,
          role: "assistant", complete: true, final: event.assistant_output.final,
          parts: event.assistant_output.parts, created_at: event.created_at,
        });
        for (const part of event.assistant_output.parts) {
          if (part.type !== "tool_call" || !part.tool_call_id) continue;
          next.tool_calls = upsertToolCall(next.tool_calls ?? [], {
            id: part.tool_call_id, conversation_id: event.conversation_id,
            run_id: event.run_id ?? "", inbound_message_id: event.inbound_message_id ?? "",
            kind: part.tool_kind ?? "", arguments: part.arguments ?? {},
            status: "requested", updated_at: event.created_at,
          });
        }
      }
      break;
    case "tool.outcome.recorded":
      if (event.tool_outcome) {
        const outcome = event.tool_outcome;
        next.messages = upsert(next.messages, {
          id: outcome.id, conversation_id: event.conversation_id,
          inbound_message_id: event.inbound_message_id, run_id: event.run_id,
          role: "tool", complete: true, created_at: event.created_at,
          parts: [{ type: "tool_result", tool_call_id: outcome.tool_call_id,
            tool_kind: outcome.tool_kind, result: outcome.result }],
        });
        next.tool_calls = (next.tool_calls ?? []).map((tool) =>
          tool.run_id === event.run_id && tool.id === outcome.tool_call_id
            ? { ...tool, status: outcome.status, result: outcome.result, updated_at: event.created_at }
            : tool);
      }
      break;
    case "run.started":
    case "run.completed":
    case "run.failed":
    case "run.stopped":
      if ((event.type === "run.failed" || event.type === "run.stopped") && event.run && event.run_id) {
        const stopped = event.type === "run.stopped";
        next.messages = upsert(next.messages, {
          id: `${event.run_id}:notice`, conversation_id: event.conversation_id,
          inbound_message_id: event.inbound_message_id, run_id: event.run_id,
          role: "notice", complete: true, created_at: event.created_at,
          notice: { status: stopped ? "stopped" : "failed", error: event.run.error },
          parts: [{ type: "text", text: stopped ? noticeStoppedText : noticeFailedText }],
        });
      }
      next.active_run = event.type === "run.started" ? event.run : undefined;
      if (event.run) {
        const submission = next.submissions?.find((item) => item.id === event.run?.inbound_message_id);
        if (submission) next.submissions = upsert(next.submissions ?? [], {
          ...submission, status: event.run.status, error: event.run.error,
        });
      }
      break;
    case "environment.progress":
      if (event.cursor && event.environment_progress &&
          !next.environment_events?.some((entry) => entry.cursor === event.cursor)) {
        next.environment_events = [...(next.environment_events ?? []), event];
      }
      break;
  }
  return next;
}
